invenio_rdm_records/services/pids/providers/datacite.py

```diff
@@ -196,16 +197,36 @@ def delete(self, pid, **kwargs):
     def validate(self, record, identifier=None, provider=None, **kwargs):
         """Validate the attributes of the identifier.
 
-        :returns: A tuple (success, errors). The first specifies if the
-                  validation was passed successfully. The second one is an
-                  array of error messages.
+        :returns: A tuple (success, errors). `success` is a bool that specifies
+                  if the validation was successful. `errors` is a list of
+                  error dicts of the form:
+                  `{"field": <field>, "messages: ["<msgA1>", ...]}`.
         """
-        _, errors = super().validate(record, identifier, provider, **kwargs)
-
-        # Format check
-        try:
-            self.client.api.check_doi(identifier)
-        except ValueError as e:
-            errors.append(str(e))
+        # success is unused, but naming it _ would interfere with lazy_gettext as _
+        success, errors = super().validate(record, identifier, provider, **kwargs)
+
+        # Validate identifier
+        # Checking if the identifier is not None is crucial because not all records at
+        # this point will have a DOI identifier (and that is fine in the case of initial
+        # creation)
+        if identifier is not None:
+            # Format check
+            try:
+                self.client.api.check_doi(identifier)
+            except ValueError as e:
+                # modifies the error in errors in-place
+                error = self._get_or_append_error_dict(errors)
```

Comment:
Nice the `check_doi`!

From a first look, I thought you had forgotten the `errors.append(error)`, but then, if I understand correctly, what is happening is:
1. `errors` is modified inside the func `_get_or_append_error_dict`, and `error` appended
2. the appended `error` is then return and modified

Wouldn't it better to pass the `message` inside and avoid this?
E.g.
```
errors = self._get_or_append_error_dict(errors, error_msg)
```
or no return value:
```
self._get_or_append_error_dict(errors, error_msg)
```
and rename the func `append_error_dict`? It looks like below that `error_msg` might a single item or a list.

Reply:
The above is indeed what was happening: returning an error_dict that had been appended so that modifications to it would modify the error_dict in `errors`.

The reason behind my approach for `get_or_append_error_dict` was just reflexive generality - the method provides a more generic functionality (retrieving an error dict guaranteed to be in given `errors`) that can be re-used in more cases and those cases are responsible for their own quirks (like appending or extending when one or multiple messages). I had been able to give a crisp name to the method (`get_or_append_error_dict` is crisp) which was a sign for me that it was a nice encapsulation. 

Anyway, I've replaced it with `_insert_pid_type_error_msg` which isn't as generic, but does the trick for what we need as you suggested.